javascript/ql/src/Statements/MissingIndexAdjustmentAfterConcurrentModification.ql

```diff
@@ -0,0 +1,163 @@
+   */
+  ArrayShiftingCall getAnArrayShiftingCall() {
+    result.getArray() = array
+  }
+
+  /**
+   * Gets a call to `splice` that removes elements from the looped-over array at the current index 
+   *
+   * The `splice` call is not guaranteed to be inside the loop body.
+   */
+  SpliceCall getACandidateSpliceCall() {
+    result = getAnArrayShiftingCall() and
+    result.getIndex().asExpr() = getIndexVariable().getAnAccess() and
+    result.getNumRemovedElements() > result.getNumInsertedElements()
+  }
+
+  /**
+   * Holds if `cfg` modifies the index variable or shifts array elements, disturbing the
+   * relationship between the array and the index variable.
+   */
+  predicate hasIndexingManipulation(ControlFlowNode cfg) {
+    cfg.(VarDef).getAVariable() = getIndexVariable() or
+    cfg = getAnArrayShiftingCall().asExpr()
+  }
+
+  /**
+   * Holds if there is a `loop entry -> cfg` path that does not involve index manipulation or a successful index equality check.
+   */
+  predicate hasPathTo(ControlFlowNode cfg) {
+    exists(getACandidateSpliceCall()) and // restrict size of predicate
+    cfg = getLoopEntry().getFirstNode()
+    or
+    hasPathTo(cfg.getAPredecessor()) and
+    getLoopEntry().dominates(cfg.getBasicBlock()) and
+    not hasIndexingManipulation(cfg) and
+
+    // Ignore splice calls guarded by an index equality check.
+    // This indicates that the index of an element is the basis for removal, not its value,
+    // which means it may be okay to skip over elements.
+    not exists (ConditionGuardNode guard, EqualityTest test | cfg = guard |
+      test = guard.getTest() and
+      test.getAnOperand() = getIndexVariable().getAnAccess() and
+      guard.getOutcome() = test.getPolarity()) and
+
+    // Block flow after inspecting an array element other than that at the current index.
+    // For example, if the splice happens after inspecting `array[i + 1]`, then the next
+    // element has already been "looked at" and so it doesn't matter if we skip it.
+    not exists (IndexExpr index | cfg = index |
+      array.flowsToExpr(index.getBase()) and
+      not index.getIndex() = getIndexVariable().getAnAccess())
+  }
+
+  /**
+   * Holds if there is a `loop entry -> splice -> cfg` path that does not involve index manipulation,
+   * other than the `splice` call.
+   */
+  predicate hasPathThrough(SpliceCall splice, ControlFlowNode cfg) {
+    splice = getACandidateSpliceCall() and
+    cfg = splice.asExpr() and
+    hasPathTo(cfg.getAPredecessor())
+    or
+    hasPathThrough(splice, cfg.getAPredecessor()) and
+    getLoopEntry().dominates(cfg.getBasicBlock()) and
+    not hasIndexingManipulation(cfg)
+  }
+}
+
+from ArrayIterationLoop loop, SpliceCall splice
+where loop.hasPathThrough(splice, loop.getUpdate().getFirstControlFlowNode())
+select splice, "Missing loop index adjustment after removing array item. Some array items will be skipped due to shifting."
```

Comment:
Could we make this alert message shorter, ideally just one sentence? Something like "Removing an array item without adjusting the loop index may cause subsequent items to be skipped" might do (and you could even incorporate the name of the loop variable into the message).